urld2fb1d2b05_thnaeff_Util_tgz-pJ8-ch_thn_util_tests_TreeTestJ8/src/ch/thn/util/thread/ControlledRunnable.java

```diff
@@ -0,0 +1,753 @@
+
+		if (pauseRequested) {
+			fireControlledRunnableListener(
+					new ControlledRunnableEvent(this, ControlledRunnableEvent.STATETYPE.PAUSED));
+		}
+
+		if (!pause && paused) {
+			//No pause any more (if currently paused)
+			synchronized (PAUSESYNC) {
+				PAUSESYNC.notify();
+			}
+		}
+
+		if (wait) {
+			if (pause && !paused || !pause && paused) {
+				//Only synchronize if really necessary
+				synchronized (WAITFORSYNC) {
+					try {
+						//Wait for entering or exiting pause state
+						while (pause && !paused || !pause && paused) {
+							WAITFORSYNC.wait();
+						}
+					} catch (InterruptedException e) {}
+				}
+			}
+
+		}
+
+	}
+
+	/**
+	 * A flag which is set when the runnable is requested to pause
+	 * (with {@link #pause(boolean)} or {@link #pause(boolean, boolean)})
+	 * 
+	 * @return
+	 */
+	protected boolean isPauseRequested() {
+		return pauseRequested;
+	}
+
+	/**
+	 * @see #runPause(boolean, boolean)
+	 */
+	protected void runPause(boolean exitAtReset) {
+		runPause(exitAtReset, false);
+	}
+
+	/**
+	 * This method implements the pausing mechanism and blocks as long as the runnable
+	 * is in the pausing state. When this method exits it could be because pausing
+	 * is done or stopping or reset has been requested. Therefore, after calling
+	 * this method there should be a check in the program flow if stopping or reset
+	 * (if implemented) has been requested.
+	 * 
+	 * @param exitAtReset If <code>true</code>, this pause is ended when {@link #reset()}
+	 * is called. If <code>false</code>, this method keeps pausing even though
+	 * the runnable is in the reset state (it ignores {@link #reset()} and
+	 * the reset state is cleared automatically).
+	 * @param controlledWait Waits even though a pause is not requested. This is
+	 * to simulate a wait().
+	 */
+	private void runPause(boolean exitAtReset, boolean controlledWait) {
+
+		if (!stopRequested && (pauseRequested || controlledWait)) {
+			//Only do the synchronization if a pause is actually requested
+
+			synchronized (PAUSESYNC) {
+				//Check the pause request again, since it could have changed
+				//just before the synchronized block. Also leave immediately if
+				//stop is requested
+				if (!controlledWait && (!pauseRequested || stopRequested)) {
+					return;
+				}
+
+				paused = true;
+				fireControlledRunnableListener(
+						new ControlledRunnableEvent(this,
+								controlledWait ? ControlledRunnableEvent.STATETYPE.WAIT : ControlledRunnableEvent.STATETYPE.PAUSED));
+
+				notifyWaitForSync();
+
+				while (!stopRequested && (pauseRequested || controlledWait)) {
+					try {
+						PAUSESYNC.wait(timeout);
+
+						//Possible timeout expired
+						if (timeout != 0) {
+							//Even though the pausing ended, do the request for ending the pause
+							//because it also notifies listeners etc.
+							pauseInternal(false, false);
+						}
+
+						if (controlledWait) {
+							break;
+						}
+					} catch (InterruptedException e) {}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)